internal/s/dyn_subtree_test.go

```diff
@@ -0,0 +1,312 @@
+		},
+	)
+}
+
+func TestDynSubtreeFailedTermination(t *testing.T) {
+	subtree := WaitDoneDynSubtree(
+		"one",
+		[]cap.Opt{},
+		[]cap.WorkerOpt{},
+		WaitDoneWorker("child0"),
+		FailTerminationWorker("child1", fmt.Errorf("child1 failed")),
+		WaitDoneWorker("child2"),
+	)
+
+	events, err := ObserveSupervisor(
+		context.TODO(),
+		"root",
+		cap.WithNodes(subtree),
+		[]cap.Opt{},
+		func(EventManager) {},
+	)
+
+	// assert that error contains all information required to assess what went wrong
+	assert.Error(t, err)
+	assert.Equal(t, "supervisor terminated with failures", err.Error())
+
+	t.Run("starts and stops routines in the correct order", func(t *testing.T) {
+		AssertExactMatch(t, events,
+			[]EventP{
+				// The dyn-subtree spawner sub-tree starts always first
+				SupervisorStarted("root/one/spawner"),
+				WorkerStarted("root/one/spawner/child0"),
+				WorkerStarted("root/one/spawner/child1"),
+				WorkerStarted("root/one/spawner/child2"),
+				WorkerStarted("root/one/worker"),
+				SupervisorStarted("root/one"),
+				SupervisorStarted("root"),
+
+				// dyn subtree terminates in reverse order
+				WorkerTerminated("root/one/worker"),
+				WorkerTerminated("root/one/spawner/child2"),
+				WorkerFailedWith("root/one/spawner/child1", "child1 failed"),
+				WorkerTerminated("root/one/spawner/child0"),
+				SupervisorFailed("root/one/spawner"),
+				SupervisorFailed("root/one"),
+				SupervisorFailed("root"),
+			})
+	})
+}
+
+func TestDynSubtreeWorkerStartStopStart(t *testing.T) {
+	subtree := cap.NewDynSubtreeWithNotifyStart(
+		"subtree",
+		func(ctx context.Context, notifyStart cap.NotifyStartFn, spawner cap.Spawner) error {
+			_, _ = spawner.Spawn(WaitDoneWorker("child0"))
+			cancelWorker, err := spawner.Spawn(WaitDoneWorker("child1"))
+			_, _ = spawner.Spawn(WaitDoneWorker("child2"))
+
+			assert.NoError(t, err)
+			err = cancelWorker()
+			assert.NoError(t, err)
+
+			cancelWorker, err = spawner.Spawn(WaitDoneWorker("child1"))
+			assert.NoError(t, err)
+			err = cancelWorker()
+			assert.NoError(t, err)
+
+			// we use notifyStart to make sure there are no race conditions in the
+			// event assertion list bellow
+			notifyStart(nil)
+
+			<-ctx.Done()
+			return nil
+		},
+		[]cap.Opt{},
+	)
+
+	events, err := ObserveSupervisor(
+		context.TODO(),
+		"root",
+		cap.WithNodes(subtree),
+		[]cap.Opt{},
+		func(EventManager) {},
+	)
+	assert.NoError(t, err)
+
+	AssertExactMatch(t, events,
+		[]EventP{
+			// The dyn-subtree spawner sub-tree starts always first
+			SupervisorStarted("root/subtree/spawner"),
+			WorkerStarted("root/subtree/spawner/child0"),
+			WorkerStarted("root/subtree/spawner/child1"),
+			WorkerStarted("root/subtree/spawner/child2"),
+			WorkerTerminated("root/subtree/spawner/child1"),
+			WorkerStarted("root/subtree/spawner/child1"),
+			WorkerTerminated("root/subtree/spawner/child1"),
+
+			WorkerStarted("root/subtree/worker"),
+			SupervisorStarted("root/subtree"),
+			SupervisorStarted("root"),
```

Comment:
It is possible to call `Spawn` after this event is emitted, correct? Can you add a test showing that?